### PR TITLE
Fixes DF-1588 - implements Browserify JS module loader

### PR DESCRIPTION
**Pull request by anselmbradford**
Refactors codebase to use Browserify for JS modules.
Fixes https://github.com/cfpb/cfgov-refresh/issues/244

The general idea is this, all code is broken across three types of modules:
 - **component-specific** =  These appear in `/src/static/js/modules/`. These work on a part of a page and are aggregated together in page-specific modules (see next).
 - **page-specific** = These appear in `/src/static/js/routes/`. These files match the URL structure of the page that they get attached to (which unfortunately means many are named `index.js`, but it has to be that way as far as I can see to keep a clean link to actual page URLs). 
 - **site-wide** = This file appears in `/src/static/js/routes/common.js`. It is added to all pages.

> **Note:** New files can be added to the routes (page-specific modules) and they will automatically be included in Browserify without any further configuration. This works through the following:
- The `browserifyPageScripts` function in the Gruntfile looks at the directory structure of the source and in addition to grabbing all the page-specific modules, generates a `routes-manifest.json` file that includes all the pages that have associated page-specific scripts.
- The routes manifest is embedded in the bottom of the base jinja template and is dynamically injected with the current page path being viewed.
- The `common.js` site-wide JS file reads the JSON from the routes manifest and if the current page path has an associated entry for a JS module it dynamically loads the module via injecting a script tag in the head of the page. This prevents all pages from requesting a page-specific module, even if they don't have one (which would throw a 404 in the console).

## Additions

- Adds `/src` directory for JS (and eventually other assets) source code.
- Adds `browserify` and dependencies.
- Adds `browserifyPageScripts` helper function to Gruntfile so that files to process can be pulled from disk [instead of being manually entered](https://github.com/cfpb/owning-a-home/blob/master/Gruntfile.js#L130-L149).
- Adds `/static/js/routes-manifest.json` to track page-specific scripts that exist.
- Adds `/modules/util/web-storage-proxy.js` utility module for handling local and session storage.
- Adds `/modules/util/js-loader.js` utility module for dynamically loading JS scripts at runtime.
- Adds additional ESLint option flags in `space-in-brackets` rule.

## Removals

- Removes inline JS from footer of homepage, `the-bureau/bureau-structure/index.html` and `the-bureau/index.html`.

## Changes

- Updates `grunt` to `~0.4.5`.
- Updates `jquery` to `^1.11.3`.
- Replaces `grunt-uglify` with `uglifyify`.
- Breaks JS into Browserify modules.
- Replaces `uglify` with `lintjs` in `jsdev` task.
- Changes ESLint indent amount to 2 spaces to match CFPB front-end standards.
- Turns off ESLint `func-names` setting because its too verbose for the gain it provides.
- General code cleanup, fixing ESLint violations (https://github.com/cfpb/cfgov-refresh/issues/300), removing extraneous whitespace, etc.

## Testing

- Throw out `/node_modules` and `/vendor` and run `npm install` and `grunt vendor`.
- Run `grunt browserify:build`.
- Check console in browser for errors.

## Review

- @jimmynotjim 
- @sebworks 
- @ascott1 
- @cfarm
- @KimberlyMunoz 

## Preview

Should be no change. If there's a change, something is broken.

## Notes

- TODO: Check if all the JS code is actually in use.
- TODO: Check if some of the global common code can move to pages.
- TODO: Add JS tests.
- `usebanner:js` in Gruntfile is broken. I'm not sure what we even use that for.
- The generated and minified JS is in `/static/js/`, once https://github.com/cfpb/cfgov-refresh/pull/466 moves forward it should be moved to `/dist/` and be added to the gitignore, because it currently contains my local path values in the source.
- I wrapped the `common.js` code in a jQuery doc ready function, testing would be helpful to determine if this isn't necessary since the code is in the footer. @Scotchester suggested there might be cases where it may still run before the DOM is ready.
- The Google Tag Manager (GTM) contains jQuery code, but loads at the top of the body, while the rest of the code (including loading jQuery) happens at the bottom. This originally was throwing errors for me but stopped when a brought the Browserify code together. The current setup of script location matches Owning a Home. However, I believe this may be hiding a bug. I think what's happening is that since the GTM is fetching its code from a CDN and the rest of the code (including jQuery) is being fetched locally, the scripts at the bottom of the body are running before the GTM has fetched its code, so when it does initialize, jQuery is available. I suspect under certain conditions it could initialize before the footer code, leading to an error of jQuery not being found.

### CODE REVIEW

**Comment by KimberlyMunoz** [comment]
Tagging in @mistergone to take a look at the Google Tag Manager issue mentioned in the notes.

File: Gruntfile.js [hunk 1]
```diff
@@ -178,23 +179,27 @@ module.exports = function(grunt) {
...
-      // headScripts: {
-      //   src: 'vendor/html5shiv/html5shiv-printshiv.js',
-      //   dest: 'static/js/html5shiv-printshiv.js'
-      // },
-      bodyScripts: {
-        src: ['static/js/main.js'],
-        dest: 'static/js/main.min.js'
+      options: {
+        // Note: The entire `browserify-shim` config is inside `package.json`.
+        transform: ['uglifyify'],
```
Comment (jimmynotjim) [inline Gruntfile.js]: Is this a typo or is it really "uglifyify"?

File: src/static/js/modules/alert.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+   Alert Initialization
+   Used on at least `/the-bureau/leadership-calendar/`.
+   ========================================================================== */
+
+'use strict';
+
+var $ = require( 'jquery' );
+
+function init() {
+	// Create the jQuery Alert plugin.
```
Comment (jimmynotjim) [inline src/static/js/modules/alert.js]: Extra indentation

File: src/static/js/modules/breakpoint-handler.js [hunk 3]
```diff
@@ -131,3 +131,9 @@ MobileOnlyExpandable.prototype.openExpandable = function () {
         this.expandableTarget.click();
     }
 };
```
Comment (jimmynotjim) [inline src/static/js/modules/breakpoint-handler.js]: Closing `;` not necessary with a declaration

File: src/static/js/modules/clear-form-buttons.js [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+   ========================================================================== */
+
+'use strict';
+
+var $ = require( 'jquery' );
+
+function init() {
+  $( '.js-form_clear' ).on( 'click', function() {
+    var $this = $( this ),
+        $form = $this.parents( 'form' );
```
Comment (jimmynotjim) [inline src/static/js/modules/clear-form-buttons.js]: Did we make a decision on declaring vars separately vs w/ comas?

File: src/static/js/modules/clear-form-buttons.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    // Clear text inputs
+    $form.find( 'input[type="text"]' ).val('');
```
Comment (jimmynotjim) [inline src/static/js/modules/clear-form-buttons.js]: The `.val` parens should include spaces

File: src/static/js/modules/jquery/cf_inputSplit.js [hunk 6]
```diff
@@ -190,3 +196,24 @@
     };
 
 })(jQuery);
```
Comment (jimmynotjim) [inline src/static/js/modules/jquery/cf_inputSplit.js]: No need to define `$` twice.

File: src/static/js/modules/jquery/custom-select.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
+'use strict';
+
+function init() {
+
+  $.fn.customSelect = function( userSettings ) {
+
+    function getSelectedOptionText($customSelect) {
```
Comment (jimmynotjim) [inline src/static/js/modules/jquery/custom-select.js]: This entire file is using 4 space indents.

File: Gruntfile.js (hunk 1 above)
Comment (jimmynotjim) [inline Gruntfile.js]: Looking at the package name, I guess it isn't ¯\_(ツ)_/¯

**Comment by jimmynotjim** [comment]
Nice work! Will be great to get the src/dist structure sorted out later so that our JS doesn't become an outlier from the CSS.

File: src/static/js/modules/alert.js (hunk 2 above)
Comment (anselmbradford) [inline src/static/js/modules/alert.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/dac47241de99a43a013ef10f38e582bb8fdcdaed

File: src/static/js/modules/jquery/cf_inputSplit.js (hunk 6 above)
Comment (anselmbradford) [inline src/static/js/modules/jquery/cf_inputSplit.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/bebb217d926085fc87d071fa3cb51ac548f9dc34

File: src/static/js/modules/breakpoint-handler.js (hunk 3 above)
Comment (anselmbradford) [inline src/static/js/modules/breakpoint-handler.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/bebb217d926085fc87d071fa3cb51ac548f9dc34

File: src/static/js/modules/clear-form-buttons.js (hunk 4 above)
Comment (anselmbradford) [inline src/static/js/modules/clear-form-buttons.js]: This is what we ended up with:

> Always declare variables in a consistent manner, and at the top of their scope. _Keeping variable declarations to one per line is encouraged._

These should be moved to single line declarations, but all these JS files need further attention. Let's fix the standards/lint issues in another PR after they are moved so it's easier to track the changes.

File: src/static/js/modules/clear-form-buttons.js (hunk 5 above)
Comment (anselmbradford) [inline src/static/js/modules/clear-form-buttons.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/bebb217d926085fc87d071fa3cb51ac548f9dc34

File: src/static/js/modules/jquery/custom-select.js (hunk 7 above)
Comment (anselmbradford) [inline src/static/js/modules/jquery/custom-select.js]: Fixed in https://github.com/cfpb/cfgov-refresh/commit/bebb217d926085fc87d071fa3cb51ac548f9dc34

File: src/static/js/modules/clear-form-buttons.js (hunk 4 above)
Comment (jimmynotjim) [inline src/static/js/modules/clear-form-buttons.js]: Agreed. Lets make that a high priority next sprint so that it's clearer what's changing in future PRs

**Comment by ascott1** [comment]
:heartbeat:

**Comment by anselmbradford** [comment]
Review edits added. Ready for squash 'n' merge?

**Comment by sebworks** [comment]
:+1:

**Comment by jimmynotjim** [comment]
:+1:

**Comment by anselmbradford** [comment]
Keep a lookout for any broken JS functionality, there was a lot moving around in here.

**Comment by anselmbradford** [comment]
FYI @KimberlyMunoz @sebworks run `grunt vendor` and `grunt browserify:build` to get your JS assets in order.

**Pull request merged**